language/move-model/src/builder/bytecode_builder.rs

```diff
@@ -0,0 +1,171 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::builder::module_builder::ModuleBuilder;
+use bytecode_source_map::source_map::SourceMap;
+use vm::{
+    file_format::{CompiledModule, Kind},
+    normalized,
+};
+
+use crate::{
+    ast::{ModuleName, QualifiedSymbol},
+    builder::model_builder::ModelBuilder,
+    model::{FunId, GlobalEnv, Loc, ModuleEnv, ModuleId, StructId},
+    symbol::Symbol,
+    ty::{PrimitiveType, Type},
+};
+
+/// Build a `GlobalEnv` from a collection of `CompiledModule`'s
+pub fn make_env(modules: Vec<CompiledModule>) -> anyhow::Result<GlobalEnv> {
+    let mut env = GlobalEnv::new();
+    let mut builder = ModelBuilder::new(&mut env);
+    for (i, m) in modules.into_iter().enumerate() {
+        let normalized = normalized::Module::new(&m);
+        let private_functions = normalized::Module::private_functions(&m);
+
+        let id = m.self_id();
+        let addr = ModuleEnv::addr_to_big_uint(id.address());
+        let module_name = ModuleName::new(addr, builder.env.symbol_pool().make(id.name().as_str()));
+        let module_id = ModuleId::new(i);
+
+        for s in normalized.structs {
+            let symbol = builder.env.symbol_pool().make(s.name.as_str());
+            let struct_name = QualifiedSymbol {
+                module_name: module_name.clone(),
+                symbol,
+            };
+            let type_params = make_type_params(&mut builder, &s.type_parameters);
+            let is_resource = !s.is_copy();
+
+            // TODO: this will crash if a struct depends on types declared in the same module.
+            // Sort normalized.structs by dependency relation so this is safe.
+            /*let mut fields = BTreeMap::new();
```

Comment:
The full builder first enters all structs and then adds their fields. Can't you do the same here? The current solution seems to be rather incomplete. Also it appears sorting would assume acyclic (non-recursive) structs. Even if that is the case right now I would not encode this as an assumption here for future resilience.

See also comment on make_type for another possible solution.